File: attractor/llm/streaming.py
"""Streaming event types and accumulator."""

from __future__ import annotations

from enum import Enum
from typing import Any

from pydantic import BaseModel, Field

from .models import (
    ContentKind,
    ContentPart,
    FinishReason,
    Message,
    Response,
    Role,
    ToolCallData,
    Usage,
)


class StreamEventType(str, Enum):
    MESSAGE_START = "message_start"
    CONTENT_DELTA = "content_delta"
    TOOL_CALL_START = "tool_call_start"
    TOOL_CALL_DELTA = "tool_call_delta"
    THINKING_DELTA = "thinking_delta"
    MESSAGE_END = "message_end"
    ERROR = "error"


class StreamEvent(BaseModel):
    type: StreamEventType
    # MessageStart fields
    id: str | None = None
    model: str | None = None
    # ContentDelta fields
    text: str | None = None
    index: int = 0
    # ToolCallStart fields
    tool_call_id: str | None = None
    tool_name: str | None = None
    # ToolCallDelta fields
    arguments_chunk: str | None = None
    # MessageEnd fields
    finish_reason: FinishReason | None = None
    usage: Usage | None = None
    # Error fields
    error_type: str | None = None
    error_message: str | None = None

    @classmethod
    def message_start(cls, id: str, model: str) -> StreamEvent:
        return cls(type=StreamEventType.MESSAGE_START, id=id, model=model)

    @classmethod
    def content_delta(cls, text: str, index: int = 0) -> StreamEvent:
        return cls(type=StreamEventType.CONTENT_DELTA, text=text, index=index)

    @classmethod
    def tool_call_start(cls, id: str, name: str, index: int = 0) -> StreamEvent:
        return cls(
            type=StreamEventType.TOOL_CALL_START,
            tool_call_id=id, tool_name=name, index=index,
        )

    @classmethod
    def tool_call_delta(cls, chunk: str, index: int = 0) -> StreamEvent:
        return cls(
            type=StreamEventType.TOOL_CALL_DELTA,
            arguments_chunk=chunk, index=index,
        )

    @classmethod
    def thinking_delta(cls, text: str, index: int = 0) -> StreamEvent:
        return cls(type=StreamEventType.THINKING_DELTA, text=text, index=index)

    @classmethod
    def message_end(
        cls, finish_reason: FinishReason, usage: Usage | None = None,
    ) -> StreamEvent:
        return cls(
            type=StreamEventType.MESSAGE_END,
            finish_reason=finish_reason, usage=usage,
        )

    @classmethod
    def error(cls, error_type: str, message: str) -> StreamEvent:
        return cls(
            type=StreamEventType.ERROR,
            error_type=error_type, error_message=message,
        )


class StreamAccumulator:
    """Collects stream deltas into a full Response."""

    def __init__(self) -> None:
        self.id: str = ""
        self.model: str = ""
        self.provider: str = ""
        self.text_parts: dict[int, list[str]] = {}
        self.tool_calls: dict[int, dict[str, Any]] = {}
        self.thinking_parts: dict[int, list[str]] = {}
        self.finish_reason: FinishReason = FinishReason.STOP
        self.usage: Usage = Usage()

    def process(self, event: StreamEvent) -> None:
        t = event.type
        if t == StreamEventType.MESSAGE_START:
            self.id = event.id or ""
            self.model = event.model or ""
        elif t == StreamEventType.CONTENT_DELTA:
            self.text_parts.setdefault(event.index, []).append(event.text or "")
        elif t == StreamEventType.TOOL_CALL_START:
            self.tool_calls[event.index] = {
                "id": event.tool_call_id or "",
                "name": event.tool_name or "",
                "arguments_chunks": [],
            }
        elif t == StreamEventType.TOOL_CALL_DELTA:
            if event.index in self.tool_calls:
                self.tool_calls[event.index]["arguments_chunks"].append(
                    event.arguments_chunk or ""
                )
        elif t == StreamEventType.THINKING_DELTA:
            self.thinking_parts.setdefault(event.index, []).append(event.text or "")
        elif t == StreamEventType.MESSAGE_END:
            self.finish_reason = event.finish_reason or FinishReason.STOP
            if event.usage:
                self.usage = event.usage

    def to_response(self) -> Response:
        parts: list[ContentPart] = []

        for idx in sorted(self.thinking_parts):
            text = "".join(self.thinking_parts[idx])
            parts.append(ContentPart.thinking_part(text))

        for idx in sorted(self.text_parts):
            text = "".join(self.text_parts[idx])
            parts.append(ContentPart.text_part(text))

        for idx in sorted(self.tool_calls):
            tc = self.tool_calls[idx]
            import json
            args_str = "".join(tc["arguments_chunks"])
            try:
                arguments = json.loads(args_str) if args_str else {}
            except json.JSONDecodeError:
                arguments = args_str
            parts.append(ContentPart.tool_call_part(tc["id"], tc["name"], arguments))

        if self.tool_calls:
            finish_reason = FinishReason.TOOL_USE
        else:
            finish_reason = self.finish_reason

        return Response(
            id=self.id,
            model=self.model,
            provider=self.provider,
            message=Message(role=Role.ASSISTANT, content=parts),
            finish_reason=finish_reason,
            usage=self.usage,
        )

File: internal/llm/streaming.go
package llm

import (
	"encoding/json"
	"sort"
)

// StreamEventType discriminates streaming events.
type StreamEventType string

const (
	StreamMessageStart  StreamEventType = "message_start"
	StreamContentDelta  StreamEventType = "content_delta"
	StreamToolCallStart StreamEventType = "tool_call_start"
	StreamToolCallDelta StreamEventType = "tool_call_delta"
	StreamThinkingDelta StreamEventType = "thinking_delta"
	StreamMessageEnd    StreamEventType = "message_end"
	StreamError         StreamEventType = "error"
)

// StreamEvent is a single event in a streaming response.
type StreamEvent struct {
	Type StreamEventType

	ID    string
	Model string

	Text  string
	Index int

	ToolCallID string
	ToolName   string

	ArgumentsChunk string

	FinishReason FinishReason
	Usage        *Usage

	ErrorType    string
	ErrorMessage string
}

// MessageStartEvent builds a message_start event.
func MessageStartEvent(id, model string) StreamEvent {
	return StreamEvent{Type: StreamMessageStart, ID: id, Model: model}
}

// ContentDeltaEvent builds a content_delta event.
func ContentDeltaEvent(text string, index int) StreamEvent {
	return StreamEvent{Type: StreamContentDelta, Text: text, Index: index}
}

// ToolCallStartEvent builds a tool_call_start event.
func ToolCallStartEvent(id, name string, index int) StreamEvent {
	return StreamEvent{Type: StreamToolCallStart, ToolCallID: id, ToolName: name, Index: index}
}

// ToolCallDeltaEvent builds a tool_call_delta event.
func ToolCallDeltaEvent(chunk string, index int) StreamEvent {
	return StreamEvent{Type: StreamToolCallDelta, ArgumentsChunk: chunk, Index: index}
}

// ThinkingDeltaEvent builds a thinking_delta event.
func ThinkingDeltaEvent(text string, index int) StreamEvent {
	return StreamEvent{Type: StreamThinkingDelta, Text: text, Index: index}
}

// MessageEndEvent builds a message_end event.
func MessageEndEvent(fr FinishReason, usage *Usage) StreamEvent {
	return StreamEvent{Type: StreamMessageEnd, FinishReason: fr, Usage: usage}
}

// ErrorEvent builds an error event.
func ErrorEvent(errType, msg string) StreamEvent {
	return StreamEvent{Type: StreamError, ErrorType: errType, ErrorMessage: msg}
}

// StreamAccumulator collects stream deltas into a full Response.
type StreamAccumulator struct {
	ID           string
	Model        string
	Provider     string
	textParts    map[int][]string
	toolCalls    map[int]*accumTool
	thinking     map[int][]string
	finishReason FinishReason
	usage        Usage
}

type accumTool struct {
	id     string
	name   string
	chunks []string
}

// NewStreamAccumulator returns an initialized accumulator.
func NewStreamAccumulator() *StreamAccumulator {
	return &StreamAccumulator{
		textParts:    map[int][]string{},
		toolCalls:    map[int]*accumTool{},
		thinking:     map[int][]string{},
		finishReason: FinishStop,
	}
}

// Process folds a single event into the accumulator.
func (a *StreamAccumulator) Process(e StreamEvent) {
	switch e.Type {
	case StreamMessageStart:
		a.ID = e.ID
		a.Model = e.Model
	case StreamContentDelta:
		a.textParts[e.Index] = append(a.textParts[e.Index], e.Text)
	case StreamToolCallStart:
		a.toolCalls[e.Index] = &accumTool{id: e.ToolCallID, name: e.ToolName}
	case StreamToolCallDelta:
		if t, ok := a.toolCalls[e.Index]; ok {
			t.chunks = append(t.chunks, e.ArgumentsChunk)
		}
	case StreamThinkingDelta:
		a.thinking[e.Index] = append(a.thinking[e.Index], e.Text)
	case StreamMessageEnd:
		a.finishReason = e.FinishReason
		if e.Usage != nil {
			a.usage = *e.Usage
		}
	}
}

// ToResponse assembles the accumulated deltas into a Response.
func (a *StreamAccumulator) ToResponse() Response {
	var parts []ContentPart

	for _, idx := range sortedKeys(a.thinking) {
		parts = append(parts, ThinkingPart(joinAll(a.thinking[idx]), "", false))
	}
	for _, idx := range sortedKeys(a.textParts) {
		parts = append(parts, TextPart(joinAll(a.textParts[idx])))
	}
	toolIdx := make([]int, 0, len(a.toolCalls))
	for k := range a.toolCalls {
		toolIdx = append(toolIdx, k)
	}
	sort.Ints(toolIdx)
	for _, idx := range toolIdx {
		tc := a.toolCalls[idx]
		argsStr := joinAll(tc.chunks)
		var args any
		if argsStr != "" {
			m := map[string]any{}
			if err := json.Unmarshal([]byte(argsStr), &m); err == nil {
				args = m
			} else {
				args = argsStr
			}
		} else {
			args = map[string]any{}
		}
		parts = append(parts, ToolCallPart(tc.id, tc.name, args))
	}

	finish := a.finishReason
	if len(a.toolCalls) > 0 {
		finish = FinishToolUse
	}

	return Response{
		ID:           a.ID,
		Model:        a.Model,
		Provider:     a.Provider,
		Message:      Message{Role: RoleAssistant, Content: parts},
		FinishReason: finish,
		Usage:        a.usage,
	}
}

func sortedKeys(m map[int][]string) []int {
	keys := make([]int, 0, len(m))
	for k := range m {
		keys = append(keys, k)
	}
	sort.Ints(keys)
	return keys
}

func joinAll(parts []string) string {
	out := ""
	for _, p := range parts {
		out += p
	}
	return out
}
